In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

In [2]:
data = pd.read_csv('EOD-APPLE.csv')
data.head()

Date    Open     High      Low   Close    Volume  Dividend  Split  \
0  28-12-2017  171.00  171.850  170.480  171.08  16480187       0.0      1   
1  27-12-2017  170.10  170.780  169.710  170.60  21498213       0.0      1   
2  26-12-2017  170.80  171.470  169.679  170.57  33185536       0.0      1   
3  22-12-2017  174.68  175.424  174.500  175.01  16349444       0.0      1   
4  21-12-2017  174.17  176.020  174.100  175.01  20949896       0.0      1   

    Adj_Open   Adj_High    Adj_Low  Adj_Close  Adj_Volume  
0  41.136972  41.341454  41.011877  41.156217    65920748  
1  40.920461  41.084047  40.826640  41.040745    85992852  
2  41.088858  41.250038  40.819183  41.033528   132742144  
3  42.022259  42.201241  41.978957  42.101646    65397776  
4  41.899570  42.344619  41.882730  42.101646    83799584

In [3]:
data.shape

(1090, 13)

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1090 entries, 0 to 1089
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Date        1090 non-null   object 
 1   Open        1090 non-null   float64
 2   High        1090 non-null   float64
 3   Low         1090 non-null   float64
 4   Close       1090 non-null   float64
 5   Volume      1090 non-null   int64  
 6   Dividend    1090 non-null   float64
 7   Split       1090 non-null   int64  
 8   Adj_Open    1090 non-null   float64
 9   Adj_High    1090 non-null   float64
 10  Adj_Low     1090 non-null   float64
 11  Adj_Close   1090 non-null   float64
 12  Adj_Volume  1090 non-null   int64  
dtypes: float64(9), int64(3), object(1)
memory usage: 110.8+ KB


In [5]:
data.describe()

Open         High          Low        Close        Volume  \
count  1090.000000  1090.000000  1090.000000  1090.000000  1.090000e+03   
mean    194.564537   196.114735   193.016676   194.613130  3.605114e+07   
std     160.562848   161.812898   159.407381   160.647536  2.194473e+07   
min      90.000000    90.700000    89.470000    90.280000  5.704900e+06   
25%     108.042500   108.942500   106.972500   108.007500  2.120714e+07   
50%     121.585000   122.504950   120.640000   121.755000  3.248325e+07   
75%     156.905000   157.809625   155.744975   156.515000  4.673778e+07   
max     649.900000   651.260000   644.470000   647.350000  1.895606e+08   

          Dividend        Split     Adj_Open     Adj_High      Adj_Low  \
count  1090.000000  1090.000000  1090.000000  1090.000000  1090.000000   
mean      0.015642     1.005505    26.367551    26.578236    26.152043   
std       0.175122     0.181735     6.614490     6.640870     6.578682   
min       0.000000     1.000000    14.189776    14.561974    14.166335   
25%       0.000000     1.000000    22.170235    22.433992    21.948836   
50%       0.000000     1.000000    25.781510    26.055842    25.511496   
75%       0.000000     1.000000    29.329099    29.541994    29.133958   
max       3.290000     7.000000    42.125703    42.628488    42.065561   

         Adj_Close    Adj_Volume  
count  1090.000000  1.090000e+03  
mean     26.373196  1.931334e+08  
std       6.616455  1.112891e+08  
min      14.258197  4.590369e+07  
25%      22.201030  1.133450e+08  
50%      25.789823  1.681963e+08  
75%      29.362853  2.394642e+08  
max      42.440845  1.065523e+09

In [6]:
data.tail()

Date    Open    High     Low    Close    Volume  Dividend  Split  \
1085  09-09-2013  505.00  507.92  503.48  506.170  12167400       0.0      1   
1086  06-09-2013  498.44  499.38  489.95  498.220  12840200       0.0      1   
1087  05-09-2013  500.25  500.68  493.64  495.270   8441700       0.0      1   
1088  04-09-2013  499.56  502.24  496.28  498.691  12322600       0.0      1   
1089  03-09-2013  493.10  500.60  487.35  488.580  11854600       0.0      1   

       Adj_Open   Adj_High    Adj_Low  Adj_Close  Adj_Volume  
1085  15.996600  16.089095  15.948451  16.033661   340687200  
1086  15.788802  15.818578  15.519869  15.781833   359525600  
1087  15.846136  15.859757  15.636755  15.688388   236367600  
1088  15.824280  15.909173  15.720381  15.796753   345032800  
1089  15.619650  15.857223  15.437510  15.476472   331928800

### Micro Task 1: How many features are there in the given data set?
#### There are 13 features in the given dataset.

In [7]:
import quandl

### Creating new features High Low Percentage(H/L_PCT) and Percentage Change(PCT_Change)

In [8]:
data['H/L_PCT'] = (data['Adj_High'] - data['Adj_Low'])/data['Adj_Close']*100.0
data['PCT_Change'] = (data['Adj_Close'] - data['Adj_Open'])/data['Adj_Open']*100.0

In [9]:
data.head()

Date    Open     High      Low   Close    Volume  Dividend  Split  \
0  28-12-2017  171.00  171.850  170.480  171.08  16480187       0.0      1   
1  27-12-2017  170.10  170.780  169.710  170.60  21498213       0.0      1   
2  26-12-2017  170.80  171.470  169.679  170.57  33185536       0.0      1   
3  22-12-2017  174.68  175.424  174.500  175.01  16349444       0.0      1   
4  21-12-2017  174.17  176.020  174.100  175.01  20949896       0.0      1   

    Adj_Open   Adj_High    Adj_Low  Adj_Close  Adj_Volume   H/L_PCT  \
0  41.136972  41.341454  41.011877  41.156217    65920748  0.800795   
1  40.920461  41.084047  40.826640  41.040745    85992852  0.627198   
2  41.088858  41.250038  40.819183  41.033528   132742144  1.050009   
3  42.022259  42.201241  41.978957  42.101646    65397776  0.527970   
4  41.899570  42.344619  41.882730  42.101646    83799584  1.097080   

   PCT_Change  
0    0.046784  
1    0.293945  
2   -0.134660  
3    0.188917  
4    0.482287

### Micro Task 2: Which symbol is used from shift method in the current scenario?

#### Symbol which is used from shift method in the current scenario is - .

In [10]:
data1 = data.copy()

In [11]:
data1 = data[['Adj_Close','PCT_Change','H/L_PCT']] 
# Take a look at the new data 
print(data1.head())

   Adj_Close  PCT_Change   H/L_PCT
0  41.156217    0.046784  0.800795
1  41.040745    0.293945  0.627198
2  41.033528   -0.134660  1.050009
3  42.101646    0.188917  0.527970
4  42.101646    0.482287  1.097080


In [12]:
# Identifying Null Values
data1.isnull().sum()

Adj_Close     0
PCT_Change    0
H/L_PCT       0
dtype: int64

As there are no NAN values in the data columns of the dataframe. There is no need to fill NAN values with -99999.

In [13]:
# A variable for predicting number of days in the future
forecast_next = 10

In [14]:
# Creating another columns i.e. Target Variable shifted 'n' units up.
data1['Predict'] = data1[['Adj_Close']].shift(-forecast_next)

<ipython-input-14-51daf262b3a3>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data1['Predict'] = data1[['Adj_Close']].shift(-forecast_next)


In [15]:
# Preview of the dataset
data1.head()

Adj_Close  PCT_Change   H/L_PCT    Predict
0  41.156217    0.046784  0.800795  41.442492
1  41.040745    0.293945  0.627198  41.305369
2  41.033528   -0.134660  1.050009  41.538719
3  42.101646    0.188917  0.527970  40.744848
4  42.101646    0.482287  1.097080  40.732819

In [16]:
data1.tail()

Adj_Close  PCT_Change   H/L_PCT  Predict
1085  16.033661    0.231683  0.877176      NaN
1086  15.781833   -0.044138  1.892738      NaN
1087  15.688388   -0.995502  1.421447      NaN
1088  15.796753   -0.173953  1.195129      NaN
1089  15.476472   -0.916650  2.711941      NaN

### Micro Task 3: What is the name of the class in which linear regression classifier is present in sklearn

#### The name of the class in which linear regression classifier is present in sklearn is LinearRegression.

In [17]:
# Creating indiependent dataset (X): converting the dataframe to a numpy array
X = np.array(data1.drop(['Predict'],1))

In [18]:
# Removing the last 10 rows 
X = X[:-forecast_next]
print(X)

[[41.15621723  0.04678363  0.80079495]
 [41.04074503  0.29394473  0.62719812]
 [41.03352802 -0.13466042  1.05000879]
 ...
 [14.96078011  0.3399193   1.3931823 ]
 [14.71940567  0.32384817  1.22449856]
 [14.42291424  1.64300377  2.68163045]]


In [19]:
# Creating indiependent dataset (y): converting the dataframe to a numpy array
y = np.array(data1['Predict'])

In [20]:
# Removing the last 10 rows 
y = y[:-forecast_next]
print(y)

[41.44249206 41.30536882 41.5387189  ... 15.68838781 15.79675289
 15.47647246]


In [21]:
# Splitting the dataset into train and test i.e. 80:20
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

In [22]:
# Model Building : linear Regression
from sklearn.linear_model import LinearRegression
lreg = LinearRegression()

In [23]:
# Model Training
lreg.fit(x_train, y_train)

LinearRegression()

In [24]:
# Model Testing: score returns the coefficient of determination R^2 of the prediction
print('Training Score: ', round(lreg.score(x_train,y_train)*100,2),'%')

# The best possible score is 1.0
print('lreg Confidence (Testing Score): ', round(lreg.score(x_test,y_test)*100,2))

Training Score:  97.37 %
lreg Confidence (Testing Score):  97.07


In [25]:
# Model Predictions

y_pred = lreg.predict(x_test)
print(y_pred)

[16.91849911 23.11157248 28.52489253 41.00704859 29.09897761 18.77779613
 25.70656473 22.41897006 16.48819694 21.76512392 37.59387441 22.10533202
 23.96640733 28.10499907 40.08956006 23.50899165 24.96135017 17.17279913
 36.81895175 27.24300539 37.83484723 24.41388451 17.2518608  21.94555751
 22.52380981 16.75153681 22.66906583 22.89022061 25.37385468 25.4737154
 28.31092429 21.27038018 24.44584497 17.33728899 24.07511752 16.30483926
 27.61428042 27.19753847 22.17875907 17.78407481 28.3881875  27.68114248
 18.96964621 25.4214168  17.60457674 25.83899302 28.98785542 26.91261215
 26.84343085 16.81951483 26.02177679 25.30125173 26.75346029 16.62013913
 27.03451587 16.08809076 25.02946517 27.21404564 28.80382257 16.94630481
 25.69310648 26.53241624 41.25327736 25.9326843  26.18925344 29.72198265
 21.74721432 27.08340355 29.56389514 28.58217503 16.46613464 34.15171469
 34.40397684 29.9003999  20.51513945 41.50531853 30.52773067 23.10255785
 37.53759856 25.05252175 26.6856373  27.11630674 28.

In [26]:
# Model Performance
from sklearn import metrics
acc=metrics.r2_score(y_test,y_pred)
print("Accuracy Score of Model: ",round(acc*100,2),'%')

Accuracy Score of Model:  97.07 %


In [27]:
# Actual v/s Predicted
prediction_df = pd.DataFrame({ 'Actual Price': y_test, 'Predicted Price': y_pred})  
prediction_df.head()

Actual Price  Predicted Price
0     17.001405        16.918499
1     24.397857        23.111572
2     28.889578        28.524893
3     37.657267        41.007049
4     29.479394        29.098978

In [28]:
# Model Evaluation
from sklearn import metrics
print('Mean Squared Error:',round(metrics.mean_squared_error(y_test,y_pred),2))
print('Root Mean Squared Error:',round(np.sqrt(metrics.mean_squared_error(y_test,y_pred)),2))

Mean Squared Error: 1.14
Root Mean Squared Error: 1.07


### Micro Task 4: What is the MSE of your model?

#### The MSE of the model is 1.14 .

In [29]:
# Forecasting
# Set x_forecast equal to the last 10 rows of the original data set from Adj. Close column
x_forecast = np.array(data1.drop(['Predict'],1))[-forecast_next:]
print(x_forecast)

[[14.25819678 -2.36008677  3.19690307]
 [14.72637449 -0.94605146  1.53349109]
 [14.97313392  0.89434371  1.98650277]
 [14.81538527  0.1498683   1.89861239]
 [15.66843166 -2.28368231  3.62890187]
 [16.03366095  0.23168318  0.87717565]
 [15.78183329 -0.04413774  1.89273815]
 [15.68838781 -0.99550228  1.42144689]
 [15.79675289 -0.17395313  1.19512884]
 [15.47647246 -0.9166498   2.71194073]]


In [30]:
# Printing lR model predictions for the next '10' days
lreg_prediction = lreg.predict(x_forecast)
print(lreg_prediction)

[15.25898141 14.89682706 14.89092815 14.86648668 16.75877895 15.72174299
 15.85182794 15.81396729 15.67528875 15.99423244]
